### PR TITLE
Add a one-directional InfoNCE loss to torch.nn

### PR DESCRIPTION
**Pull request by GiacomoSaccaggi**
## Summary

Adds a one-directional InfoNCE loss to `torch.nn`, addressing #97539.

Closes #97539

`positive_key[i]` is the positive for `query[i]`; pairs are aligned by index and
embeddings are L2-normalized internally. Two negative-sampling modes are
supported:

- **in-batch negatives** (`negative_keys=None`): the `(N, N)` cross-view
  similarity matrix, whose diagonal holds the positive pairs, so the target for
  row `i` is `i`;
- **explicit negatives**: a shared set of `M` keys scored against every query,
  which is the softmax classification form used by
  [MoCo](https://arxiv.org/abs/1911.05722).

Both paths build logits and delegate to `F.cross_entropy`, so numerical
stability comes from the existing fused log-softmax kernel rather than a
hand-rolled softmax.

```python
loss = F.info_nce_loss(query, positive_key)                 # in-batch negatives
loss = F.info_nce_loss(query, positive_key, negative_keys)  # shared explicit negatives

loss_fn = nn.InfoNCELoss(temperature=0.07)
loss = loss_fn(query, positive_key)
```

`none`, `mean` and `sum` reductions are supported.

## Scope and semantics

The objective is deliberately **one-directional**, and this is not the exact
NT-Xent objective of [SimCLR](https://arxiv.org/abs/2002.05709): SimCLR treats
both views as anchors over a shared set of `2N` representations. Averaging both
directions of this loss gives a CLIP-style symmetric objective with a different
denominator, so the docstrings say this explicitly instead of claiming
equivalence:

```python
loss = 0.5 * (
    F.info_nce_loss(query, positive_key)
    + F.info_nce_loss(positive_key, query)
)
```

The loss owns no state: no memory queue, no momentum encoder, no stop-gradient,
no data augmentation. A MoCo-style caller passes its external queue as
`negative_keys` and decides outside the loss whether the keys receive gradients;
nothing is detached internally.

Two API surfaces were considered and left out of this PR: per-query negatives of
shape `(N, M, D)`, and a `symmetric=True` flag. Both widen the public API and can
be revisited once the base semantics are settled.

A `SupConLoss` implementation was part of an earlier revision of this PR. It has
been removed so that the API and semantics of InfoNCE can be reviewed on their
own; it will be proposed separately with its own issue.

## Files changed

| File | Lines | Description |
| --- | --- | --- |
| `torch/nn/functional.py` | +152 | `info_nce_loss()` |
| `torch/nn/modules/loss.py` | +110 | `InfoNCELoss` module (thin wrapper) |
| `torch/nn/modules/__init__.py` | +2 | Import and `__all__` export |
| `torch/nn/functional.pyi.in` | +10 | Type stub |
| `torch/overrides.py` | +3 | `__torch_function__` testing override |
| `docs/source/nn.md` | +1 | Sphinx autodoc entry |
| `docs/source/nn.functional.md` | +1 | Sphinx autodoc entry |
| `CODEOWNERS` | +1 | Register the new test file |
| `test/test_info_nce_loss.py` | +513 | 52 tests |

Suggested review order: `torch/nn/functional.py` for the primitive and its
contract, `torch/nn/modules/loss.py` for the module wrapper, then the
registration files, and finally the tests.

## Testing

```
python test/test_info_nce_loss.py
python test/test_overrides.py -k info_nce
python test/test_overrides.py
python test/test_public_bindings.py
```

52 tests pass, covering both negative modes against an independent
`F.cross_entropy` reference, all three reductions, `gradcheck` and
`gradgradcheck`, gradients through every input including `negative_keys`,
batch-size-one returning a zero that is still connected to the graph, invalid
shapes, invalid and non-finite temperatures, `__torch_function__` dispatch on
each tensor argument in turn, TorchScript, and `torch.compile(fullgraph=True)`.

The full `test_overrides.py` run (7705 tests) confirms the override registration
is consistent, and `test_public_bindings.py` confirms the public surface.
Coverage was measured on CPU; my machine has no CUDA, so the CUDA path is left
to CI.

## Open questions for reviewers

Three points where a maintainer decision would help:

- the default `temperature=0.07` follows MoCo, while SimCLR-style setups
  typically use 0.1 to 0.5. Is 0.07 the right default, or should the argument
  have none?
- embeddings are always L2-normalized internally. Some callers arrive with
  pre-normalized embeddings; would you prefer that be assumed instead?
- the temperature check rejects NaN and infinity. No neighbouring loss validates
  the finiteness of its scalar arguments, so this is slightly stricter than local
  convention. I kept it because a non-finite temperature otherwise yields a
  silently NaN loss.

I am happy to reduce the surface to the functional form only if that is
preferred.

This PR was prepared with the assistance of an AI coding assistant.

Co-authored-by: Elisa Tedde <elitedde1@gmail.com>

### CODE REVIEW

**Comment by GiacomoSaccaggi** [comment]
I ran the InfoNCE test suite locally on CPU, as my pc does not have CUDA GPU.

All 52 tests in `test/test_info_nce_loss.py` pass, along with the `__torch_function__` override test in `test/test_overrides.py`.

Test coverage includes:

* Comparisons against `F.cross_entropy` for both in-batch and explicit-negative paths
* All three reduction modes
* `gradcheck` and `gradgradcheck`
* The batch-size-one case, which returns a connected zero
* Validation of invalid shapes and non-finite temperatures
* `__torch_function__` dispatch for each tensor argument
* TorchScript
* `torch.compile(fullgraph=True)`

The complete `test/test_overrides.py` suite, comprising 7,705 tests, also passes locally. The CUDA path is left for CI validation.


image

**Comment by elitedde** [comment]
Following up on our earlier comment where we could only test on CPU — I've now run the full test suite on a CUDA machine:

**Environment**

* **OS & Python:** Windows 10, Python 3.12.10
* **GPU:** NVIDIA GeForce GTX 1050 (4 GB, Compute Capability 6.1, Driver 528.79)
* **PyTorch:** `torch 2.7.1+cu118` with this PR's changes applied directly to the installed wheel (`torch/nn/functional.py`, `torch/nn/modules/loss.py`, `torch/nn/modules/__init__.py`, `torch/overrides.py`) since building from source wasn't feasible on this machine and the PR is pure Python.

**Results**

```bash
python test/test_info_nce_loss.py
# Ran 55 tests in 4.7s
# OK (skipped=1)

```

With a GPU available, `instantiate_device_type_tests` now generates and runs the CUDA variants, all passing:

* `test_device_of_output_matches_input_cuda`
* `test_gradient_on_device_cuda`
* `test_cpu_double_precision_cuda` (skipped as expected via `@onlyCPU`)

The remaining 52 tests (`gradcheck`/`gradgradcheck`, `torchscript`, `compile`, `dtypes`, edge cases) all pass, matching the CPU results.

**Additional CUDA Sanity Checks**

* **CPU vs. CUDA Parity:** `fp32` results are bit-identical on identical inputs (including with explicit `negative_keys`).
* **Precision Stability:** `fp16` and `bfloat16` on CUDA produce finite losses (no NaN/inf).
* **Gradcheck:** `torch.autograd.gradcheck` passes in double precision on CUDA.


image

**Comment by GiacomoSaccaggi** [comment]
Rebased on latest main, all tests passing locally. @drisspg when you have a moment, could you kick off CI? 
Happy to address any feedback. Thanks!

**Comment by GiacomoSaccaggi** [comment]
Hi everyone! I've updated the branch with the latest changes from main and rerun the local tests, which are passing.

Could a maintainer please approve the CI workflows when they have a moment? @albanD  and @jbschlosser , I'd also really appreciate your feedback on the PR whenever you have time. Happy to make any adjustments needed.

Thanks for your time and help!
